### ISSUE TITLE
Apple M1

### ISSUE THREAD

**Issue by weaverm** [donmelton/other_video_transcoding#86]
Do folks here have any thoughts or experience transcoding with the M1?  I see that Handbrake has a Universal Binary beta ([HandBrake 1.4.0-beta.1](https://github.com/HandBrake/HandBrake/releases/tag/1.4.0-beta.1)). I have no idea where ffmpeg is on the Universal Binary spectrum. I also am not aware of what, if any, changes have been made to the Apple VideoToolBox in Big Sur and on the M1.

**Comment by samhutchins** [comment]
There's no technical reason things wouldn't work, all the stuff required for `other-transcode` can be compiled for ARM (I've been running it on a Raspberry Pi recently)

Not sure what HomeBrew's plans for M1 are, but once that's serving ARM binaries (Universal or otherwise), everything should work.

I'd expect VideoToolbox to be similar to T2 macs, but I'm honestly not sure what changes have been made to the encoding block since T2. I doubt it's any worse.

I'd expect all the software encoders to work, and probably faster than most other macs

**Comment by lambdan** [comment]
I got the M1 Air (8/8 cores, 8 ram, 512 ssd) and I ran The Godfather (1080p bluray) through it using the VideoToolbox HEVC encoder (all default settings otherwise, just used the `--hevc` switch):

     > ask-ffmpeg-log The\ Godfather_t00.mkv.log 
     > 191 fps 00:22:14 The Godfather_t00.mkv

Pretty good I think? Especially since this thing doesn't even have a damn fan in it. I

As you can see in this screenshot, ffmpeg runs through Rosetta ("Intel Architecture") but VTEncoderXPCService is running under the "Apple architecture". 

Video quality wise I think its fine? I can't tell see any obvious flaws in it. 

![Screenshot 2020-11-28 at 14 50 21](https://user-images.githubusercontent.com/1690265/100517293-ea881700-3189-11eb-903a-ecf550ade1da.png)

**Comment by donmelton** [comment]
@lambdan Thanks for testing and sharing these numbers!

One of our HiveMind™ members is seeing similar performance on her new M1 MacBook Pro with a still-Intel `ffmpeg` binary for HEVC and AVC/H.264 output.

**Comment by mushbuddygoose** [comment]
I am seeing same thing. I found a static FFmpeg compiled for Apple Silicon but doesn't seem to make a difference. Using the VideoToolBox, 1080P BR > H265 at CB of 8K yields 200 FPS on my Mac mini M1. When I drop it down to x265 at medium and CF 20 I am seeing 14/15 FPS!

**Comment by donmelton** [comment]
@mushbuddygoose Thanks for sharing those numbers!

**Comment by weaverm** [comment]
Can anyone comment on if VideoToolBox on M1 can produce 10-bit HEVC output from an 8-bit input?

**Comment by donmelton** [comment]
@weaverm The HiveMind™ has been doing some testing on this and it appears that current M1-equipped Macs cannot produce 10-bit HEVC output using VideoToolbox. However, we don't know if this is a limit in the hardware, in system software or in `ffmpeg`. If it's a software issue then hopefully it can be fixed in the future.

**Comment by RusselProuse34** [comment]
I also hope this is something that can be fixed because the M1 Mac mini seems like the ultimate machine to run a media server on - cool, silent, low power, and incredibly fast.

**Comment by donmelton** [comment]
@RusselProuse34 This is exactly why the nerds of the HiveMind™ are interested in it. :)

**Comment by weaverm** [comment]
I got an M1 Air a couple days ago. Homebrew installed arm64_big_sur versions of everything for FFmpeg and MKVToolNix. When I run other-transcode on it, ffmpeg shows up as Apple architecture in Activity Monitor. The --vt encoder works as expected (still only 8-bit hevc from 8-bit source). The --x265 encoder will make 10-bit output, but it is slow. Roughly 4 fps with a 1080p blu-ray rip and a --target 4000 encode. 

```
Transcoding...
x265 [info]: HEVC encoder version 0.0
x265 [info]: build info [Mac OS X][clang 12.0.0][32 bit][noasm] 10bit
x265 [info]: using cpu capabilities: none!
x265 [info]: Main 10 profile, Level-4 (Main tier)
x265 [info]: Thread pool created using 8 threads
```

I did find it humorous that the cpu capabilities line with --x265 ends with an exclamation point.

**Comment by donmelton** [comment]
@weaverm Thanks for testing! 👍 

Yes, using `--x265` will always be the slowest option. 🤷‍♂️

**Comment by Thor263** [comment]
I'm getting an M1 mini next week, so I've been following this with some interest too. It appears that x265 hasn't been optimized for M1 yet -- see issue [#578](https://bitbucket.org/multicoreware/x265_git/issues/578/implement-arm64-compiling-for-apple) at [x265_git](https://bitbucket.org/multicoreware/x265_git/wiki/Home). However, looking at that issue, it appears that Handbrake has implemented much more complete x265 optimizations, so hopefully that will be coming back to main x265 soon.

**Comment by clunkclunk** [comment]
I just did a transcode with other-transcode on my M1 Mac Mini, and it went well. Installed ffmpeg and mkvtoolnix via homebrew (which is recently installed, and mostly working for ARM/M1). mpv wouldn't install because it's not ready for arm, but I didn't bother to dig in further.

Used --hevc and it appears to be leveraging the VTEncoder, though I only have 8 bit files to test, rather than 10 bit. Speeds seem to be around 200-225fps for a 1080p x264 source.

Oh and these M1 Macs are amazing. This is a base model, and I'm also simultaneously using Slack, Mail, Chrome and a Google Meet in Chrome with zero feelings of lag or stuttering, and no fan noise.

**Comment by Skywalka1976** [comment]
You can use Homebrew's x86 version by using

arch -x86_64

As a prefix before all commands to make everything work. And yes the transcode speeds are amazing on the M1.

**Comment by Botts85** [comment]
I've been playing with the M1 and Handbrake / ff-works quite a bit.  It's lightning fast, it crushes my i9 iMac at x265 hardware encodes at 1080p and 4K.

It looks like the M1's videotoolbox support was tuned for SSIM.  Irrespective of bitrate, it beats x265 slow and 9th gen QuickSync based VTB.  

However, it loses in VMAF against both QuickSync based VTB and X265 slow.

**Please ignore the color explosion.  Teal is M1 hardware.  Green is i9 iMac hardware encoding.**

Source data is here: https://docs.google.com/spreadsheets/d/1-lq8mA1z5srjBLJysdiUEN90HaCkuW7g_FPyyCJVenQ/edit?usp=sharing

<img width="1204" alt="Screen Shot 2021-02-14 at 3 56 20 PM" src="https://user-images.githubusercontent.com/13722323/107891658-476bb180-6edd-11eb-99a6-a8fa834da3cf.png">

**Comment by donmelton** [comment]
@Botts85 Interesting. Thanks for posting these numbers.

Personally, I don't put much stock in PSNR, SSIM or VMAF being any kind of absolute measure of quality, objective or not. Those metrics are, of course, useful, but nothing beats a pair of Mark I™ eyeballs for assessing quality. 😄 

Especially when similar target bitrates can produce significant differences between encoders. And things can get very different once you mix in presets.

So my question is, which one looks better to you?

**Comment by Botts85** [comment]
The Mark 1 eyeball agrees with VMAF in this case.  X265 slow still wins to my eyes.  Then Quick Sync VideoToolbox, then M1's VideoToolbox.

It was a sad day when I tested M1 hoping it'd bring Apple's "magic" to my encode life and noticed it was worse than my Intel Mac for quality.

I haven't tested M1's CQ H265 encoding too much yet, but it doesn't seem massively better.

For "production" work, I'd still use software.  For streaming though, M1's VTB is insanely fast.  The challenge I'm looking at is when backing up my bluray library, is it worth the time of X265 slow, or do I just throw extra bitrate at it and use VTB.

I'm actually trying to figure out a way to test if the M1's video quality impacts Final Cut Pro's workflow in any notable way versus a Mac with Quick Sync, or just CPU encoding.

**Comment by donmelton** [comment]
@Botts85 As far as backing up your Blu-ray library goes, a remux (not a transcoding) or your original rip is your best bet. Transcoding is really for making things smaller and more portable. But when you have enough storage space, decent bandwidth and a Plex server, then viewing your originals is preferred.

Seriously, investing in more storage will be cheaper in the long run than waiting for `x265` to deliver you the "perfect" transcode.

And if you don't have enough storage to leverage that strategy, then I would just use a hardware transcoder, even the M1's VT technology, to get smaller and portable versions of your videos. They'll be "good enough," especially if you can get 10-bit HEVC output.

My two cents. 😄

**Comment by Botts85** [comment]
Totally fair point on that.  I think I've got 72TB of storage in my rack right now, but was hoping to avoid too much expansion.

I'd been running Jellyfin for on the fly transcode when over WAN.  So far, I've liked it more than Plex and it's free, but you'd need to set up your own VPN or reverse proxy for it.

I can always re-rip, so good point with just using the M1 for now, then later on when storage permits go either to remux'd versions, or maybe rip to AV1 if we ever get a faster encoder for it.

**Comment by RusselProuse34** [comment]
As per [this comment](https://forums.macrumors.com/threads/hevc-10-bit-compressor-hw-encode.2277539/?post=29447907#post-29447907) from the MacRumors forums, it appears that someone was able to get 4K content encoded in 10-bit using VT in Compressor back in December. I'm not an expert by any means, but if Compressor can do it is it possible that ffmpeg could also add this functionality?

**Comment by donmelton** [comment]
@RusselProuse34 It's unclear since Compressor is Apple's proprietary software. Which means they have access to APIs and mechanisms that FFmpeg might not.

**Comment by Botts85** [comment]
@RusselProuse34 Handbrake's newest nightly will do 10-bit H265 (non-HDR IIRC) VideoToolbox encoding.

There'a an ffmpeg commit here: https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210122202133.31519-1-simone@lisanet.de/ that should address it in ffmpeg.

` Encoder: Apple HEVC (HW)
    Codec Type: HEVC/H.265
    Encoder ID: com.apple.videotoolbox.videoencoder.ave.hevc
    Codec Name: HEVC
    Encoder Name: Apple HEVC (HW)
    Supported Properties:
        iChatUsageString
            Value is read-write.
        AverageBitRate
            Value is read-write.
        DataRateLimits
            Value is read-write.
        TransferFunction
            Value is read-write.
            ITU_R_709_2
            SMPTE_240M_1995
            Linear
            IEC_sRGB
            ITU_R_2020
            SMPTE_ST_2084_PQ
            SMPTE_ST_428_1
            ITU_R_2100_HLG
            UseGamma
        FrameRateTargetForAverageBitrate
            Value is read-write.
        PoolPixelBufferAttributesSeed
        FigThreadPriority
        MaxKeyFrameIntervalDuration
            Value is read-write.
        UserRPSForFaceTime
            Value is read-write.
        ChromaLocationTopField
        RequireDeterministicDependencyAndReordering
            Value is read-write.
        ForceSliceRPS
            Value is read-write.
        MinimizeMemoryUsage
            Value is read-write.
        FieldDetail
        EnableMBInputCtrl
            Value is read-write.
        ProfileLevel
            Value is read-write.
            HEVC_Main_AutoLevel
            **HEVC_Main10_AutoLevel**
            HEVC_MainStill_AutoLevel
            HEVC_Main444_AutoLevel
            **HEVC_Main44410_AutoLevel**
            **HEVC_Main42210_AutoLevel**
            HEVC_Monochrome12_AutoLevel
            HEVC_Monochrome_AutoLevel
        EnableContentColourVolumeSEI
            Value is read-write.
        UsingHardwareAcceleratedVideoEncoder
            Value is read-write.
        PixelBufferPoolIsShared
        EnableWeightedPrediction
            Value is read-write.
        InputPixelFormat
            Value is read-write.
        YCbCrMatrix
            Value is read-write.
            ITU_R_709_2
            ITU_R_601_4
            SMPTE_240M_1995
            ITU_R_2020
        InputQueueMaxCount
            Value is read-write.
        SoftMinQuantizationParameter
            Value is read-write.
        ComponentRange
        AllowOpenGOP
            Value is read-write.
        ColorPrimaries
            Value is read-write.
            ITU_R_709_2
            EBU_3213
            SMPTE_C
            **ITU_R_2020**
            P3_D65
        HEVCTemporalLevelInfoCombinations
            Value is read-only.
        AllowTemporalCompression
            Value is read-write.
        AllocCodedBufferHalfSize
            Value is read-write.
        GammaLevel
        ICCProfile
        SoftMaxQuantizationParameter
            Value is read-write.
        ForceRefUncompressed
            Value is read-write.
        EnableUserRefForFacetime
            Value is read-write.
        InsertTrailingBytes
            Value is read-write.
        PoolPixelBufferAttributes
        HDRMetadataInsertionMode
        RealTime
            Value is read-write.
        UserParameterSetsIds
            Value is read-write.
        RequiredHEVCPictureParameterSetID
            Value is read-write.
        RequiredHEVCSequenceParameterSetID
            Value is read-write.
        DebugMetadataSEI
            Value is read-write.
        MaxQuantizationParameter
            Value is read-write.
        LossLess
            Value is read-write.
        CleanAperture
        DroppablePFramesOnly
            Value is read-write.
        UseVideoResolutionAdaptation
            Value is read-write.
        EnableParallelStitching
            Value is read-write.
        EnableStatsCollect
            Value is read-write.
        ExpectedFrameRate
            Value is read-write.
        PerceptualQualityOptimization
            Value is read-write.
        NumberOfPendingFrames
        OutputBitDepth
            Value is read-write.
        DebugMetadataSEIuseEPB
            Value is read-write.
        log2_max_minus4
            Value is read-write.
        AmbientViewingEnvironment
            Value is read-write.
        PixelAspectRatio
        ThrottleForBackground
            Value is read-write.
        Priority
            Value is read-write.
        ConvergenceDurationForAverageDataRate
            Value is read-write.
        PreserveDynamicHDRMetadata
        AuxiliaryTypeInfo
        EncodesAlpha
            Value is read-write.
        ExpectedDuration
            Value is read-write.
        MultiReferencePSpacing
            Value is read-write.
        UserDPBFramesForFaceTime
            Value is read-write.
        HEVCSyncSampleNALUnitTypes
            Value is read-only.
        Quality
            Value is read-write.
        MultiPassStorage
        EnableTimingInfoMetadata
            Value is read-write.
        EncoderUsage
            Value is read-write.
        PixelTransferProperties
        UsingMetalRegistryID
        ChromaLocationBottomField
        FieldCount
        AllowFrameReordering
            Value is read-write.
        RequiredHEVCVideoParameterSetAlphaLayerID
            Value is read-write.
        EncoderID
        AllowPixelTransfer
        VideoEncoderPixelBufferAttributes
        DeblockingFiltering
            Value is read-write.
        ContentLightLevelInfo
        ChromaQPIndexOffsetMultiPPS
            Value is read-write.
        EnableMultiReferenceP
            Value is read-write.
        MaxKeyFrameInterval
            Value is read-write.
        BaseLayerFrameRate
            Value is read-write.
        EnableUserQPForFacetime
            Value is read-write.
        MaximizePowerEfficiency
            Value is read-write.
        PrioritizeEncodingSpeedOverQuality
            Value is read-write.
        MasteringDisplayColorVolume
        QuantizationScalingMatrixPreset
            Value is read-write.
        ClientPID
        SAOMode
            Value is read-write.
        SourceFrameCount
            Value is read-write.
        CodedBuffersOverride
            Value is read-write.
        RelaxAverageBitRateTarget
            Value is read-write.
        NumberOfSlices
            Value is read-write.`

This is what my M1 MacBook Pro lists for support for HEVC with VideoToolbox.  Some emphasis added.

**Comment by donmelton** [comment]
@Botts85 Thanks for the information. 👍

**Comment by Botts85** [comment]
I got a chance to play with an M1 Pro today.  It's definitely fast with VideoToolbox, but sadly, the encoder for H265 looks identical to the M1.

Aka, it is eclipsed by 9th gen QuickSync and NVENC.

I need to do some more testing, but it seems like the "CQ" setting that Handbrake has for M1 VTB is just a slider for a set of different CBRs.  Take a sample video file in 4K, encode one copy at 4K CQ38, then encode a second copy at 1080P CQ38 and it'll likely wind up exactly the same bit rate.  That's unfortunate.

**Comment by donmelton** [comment]
@Botts85 Thanks for the update! That is... disappointing but predictable, I suppose.

**Comment by tylercheung** [comment]
I would agree with your eyeballs.  Maybe placebo but I am noticing more pixelation especially on older films with out-of-focus lens bokeh and shadows… i’ve re-transcoded on my linux box w nvenc and a 1080.    I haven’t figured out how to set things up on my windows box yet, was going to try chocolatey to install ffmpeg next

**Comment by donmelton** [comment]
@tylercheung Rather than use Chocolately to install FFmpeg, try this:

https://github.com/donmelton/other_video_transcoding/wiki/Windows

**Comment by tylercheung** [comment]
> @tylercheung Rather than use Chocolately to install FFmpeg, try this:
> 
> 
> 
> https://github.com/donmelton/other_video_transcoding/wiki/Windows
> 
> 

Ah...kind of a tangent but in troubleshooting my windows install, perusing my windows path, I discovered a third party video conferencing app ("Intouch Health/Teladoc") used for work) had its own conflicting ffmpeg install tucked away somewhere...go figure!

**Comment by sammoth** [comment]
You can use the constant quality mode of videotoolbox encoders (HEVC and x264) on Apple Silicon from ffmpeg using `-q:v` and I do not find that it just translates to a target bitrate - encoding the same clip at 4k and 2k with the same q value produces completely different filesizes.

**Comment by donmelton** [comment]
@sammoth Yep. I would expect a constant quality mode to do just that, i.e. produce variable-sized output.
